contents/kotlin/NullSafety.md

```diff
@@ -0,0 +1,142 @@
+
+```java
+String a = null;
+if(a.length > 5) {
+    //do something
+}
+```
+When the above code is ran, a NPE will be thrown on line 2 because a `null` object has no methods. To prevent an object from taking on a `null` value, programmers typically resort to doing additional checks like this:
+```java
+String a = null;
+if(a != null && a.length > 5) {
+        //do something
+}
+```
+And of course, that's all fine, until we want to do something more complex. Say Bob belongs to a department, and we want to get the name of the department manager. That will look like this:
+```java
+String managerName = bob.department.manager.name;
+```
+
+Because each variable can be `null`, to prevent the NPE we put the code in the following code block:
+```java
+if(bob != null) {
+    Department department = bob.department;
+    if(department != null) {
+        Employee manager = department.manager;
+        if(manager != null) {
+            String name = manager.name;
+            if(name != null) {
+                //do something
+            }
+        }
+    }
+}
+```
+The deep-nested `if` statement adds to the verbosity of our code.
+
+The other way is to use [Java Optionals](http://www.oracle.com/technetwork/articles/java/java8-optional-2175753.html).
+For the first example above, we can do
+``` java
+a.ifPresent(this::doSomething);
+```
+
+And for the deeply-nested `if` statements, the verbosity can be reduced with
+``` java
+bob.map(Person::getDepartment)
+    .map(Person::getManager)
+    .map(Person::getName)
+    .ifPresent(Person::doSomething);
+```
+
+Let's see how Kotlin deals with this issue while maintaining a simple and readable syntax.
+
+# How does Kotlin handle Null Safety?
+## Nullable and Non-nullable type
+In Kotlin, a type can be nullable or non-nullable, determined by the presence of a `?`. For example, an object of type `String` is non-nullable, while an object of type `String?` is nullable.  
+
+As the compiler catches `null` assignments to non-nullable objects, the following would result in compilation error.
+```Kotlin
+var firstString: String = "hello world"
+firstString = null //compilation error
+```
+In comparison, the following assignment to a nullable type is allowed
+```Kotlin
+var secondString: String? = "hello world"
+secondString = null //okay
+```
+
+In the first case, we can safely call `firstString.length` without having to worry about a NPE because `firstString` can never be `null`.   
+In the second case, `secondString` can potentially be `null`, so `secondString.length` will result in a compilation error as the compiler see the danger of such statement and blocks it early.
+
+
+## Operators in Kotlin
```

Comment:
This heading is too general. `Operators used for Null Safety` or something like that?